src/data/dataloader.cpp

```diff
@@ -0,0 +1,693 @@
+                                LoadDataFunc<T> loadFunction,
+                                LoadCacheFunc<T> loadCacheFunction,
+                                SaveCacheFunc<T> saveCacheFunction)
+    {
+        static_assert(
+            std::is_same_v<T, openspace::dataloader::Dataset> ||
+            std::is_same_v<T, openspace::dataloader::Labelset> ||
+            std::is_same_v<T, openspace::dataloader::ColorMap>
+        );
+
+        std::string info;
+        if (specs.has_value()) {
+            info = openspace::dataloader::generateHashString(*specs);
+        }
+        std::filesystem::path cached = FileSys.cacheManager()->cachedFilename(filePath, info);
+
+        if (std::filesystem::exists(cached)) {
+            LINFOC(
+                "DataLoader",
+                fmt::format("Cached file {} used for file {}", cached, filePath)
+            );
+
+            std::optional<T> dataset = loadCacheFunction(cached);
+            if (dataset.has_value()) {
+                // We could load the cache file and we are now done with this
+                return *dataset;
+            }
+            else {
+                FileSys.cacheManager()->removeCacheFile(cached);
+            }
+        }
+        LINFOC("DataLoader", fmt::format("Loading file {}", filePath));
+        T dataset = loadFunction(filePath, specs);
+
+        if (!dataset.entries.empty()) {
+            LINFOC("DataLoader", "Saving cache");
+            saveCacheFunction(dataset, cached);
+        }
+
+        return dataset;
+    }
+} // namespace
+
+namespace openspace::dataloader {
+
+namespace data {
+
+Dataset loadFile(std::filesystem::path path, std::optional<DataMapping> specs) {
+    ghoul_assert(std::filesystem::exists(path), "File must exist");
+
+    std::ifstream file(path);
+    if (!file.good()) {
+        throw ghoul::RuntimeError(fmt::format("Failed to open data file {}", path));
+    }
+
+    std::string extension = path.extension().string();
+    for (size_t i = 0; i < extension.size(); i++) {
+        extension[i] = static_cast<char>(tolower(extension[i]));
+    }
+
+    Dataset res;
+    if (extension == ".csv") {
+        res = csv::loadCsvFile(path.string(), specs);
+    }
+    else if (extension == ".speck") {
+        res = speck::loadSpeckFile(path.string(), specs);
+    }
+    else {
+        LERRORC("DataLoader", fmt::format(
+            "Could not read data file {}. File format {} is not supported",
+            path, path.extension()
+        ));
+    }
+
+    return res;
+}
+
+std::optional<Dataset> loadCachedFile(std::filesystem::path path) {
+    std::ifstream file(path, std::ios::binary);
+    if (!file.good()) {
+        return std::nullopt;
+    }
+
+    Dataset result;
+
+    int8_t fileVersion;
+    file.read(reinterpret_cast<char*>(&fileVersion), sizeof(int8_t));
+    if (fileVersion != DataCacheFileVersion) {
+        // Incompatible version and we won't be able to read the file
```

Comment:
Nah, I don't think so.  From his/her side it doesnt matter if the cache version of the file works or not, and it will result in a new file being cached. 

In general, the user should not have to know/think/care about the things in the cache folder (unless absolutely necessary, I guess)